### PR TITLE
[feature] JEPAs base trainer

### PR DESCRIPTION
**Pull request by sami-bg**
## Description

Implements a new base trainer for Joint Embedding Predictive Architectures (JEPAs). JEPAs, in the broadest terms, consist of:
1. A target input signal is transformed into a context input signal with some parameters `a`.
2. A target encoder that encodes an input signal into a representation.
3. A predictor predicts the encoder's representation of the context signal, given the target representation and the transformation parameters `a`.

image
image taken from [Image World Models](https://arxiv.org/pdf/2403.00504)

In the example of I-JEPA:
1. A target image is masked (this is the transformation) into a context with some masking positions (these are the parameters of the transformation).
2. A target encoder encodes this image into patches.
3. A context encoder encodes the masked image into patches. This context encoder is trained and the target encoder's parameters are frozen and updated with slow-moving EMA to prevent collapse.
4. A predictor is given the masking positions that produced the context, as well as a patch location, and predicts the representation of the target patch in that location. In I-JEPA, the predictor predicts 4 target blocks, one after the other.

V-JEPA follows a very similar training procedure.

This is still a WIP. I intend to add I-JEPA and V-JEPA to StableSSL, but not sure if:
1. Does the scope of stableSSL cover specific architectures?
2. If so, do they belong in this PR? 
3. They will involve some data collation functions for the dataloader to easily condition the predictor on the transformation parameters. I have a few ideas of where in the repo it should be (maybe its own `collators.py` file). From the i-jepa and v-jepa repos, they seem to need quite elaborate implementations.

## Checklist

- [x] I have read the [**Contributing**](https://rbalestr-lab.github.io/stable-SSL.github.io/dev/contributing.html#) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [ ] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR to the [**RELEASES.rst**](../RELEASES.rst) file.

### CODE REVIEW

**Comment by sami-bg** [comment]
Howdy! 

I tested the JEPA base trainer using [this](https://github.com/sami-bg/v-jepa-distractors) repo, which fully integrates stable-ssl with the [jepa code](https://github.com/facebookresearch/jepa) from FAIR (with several modifications). It subclasses the JEPA base trainer into VJEPA_Trainer, and includes the config file I used to run pre-training [here](https://github.com/sami-bg/v-jepa-distractors/blob/main/configs/pretrain/vjepa/test_config.yaml). I granted you both access in case you'd like to gloss it over, but pre-training seems to work well! 

I made sure to keep the JEPA base trainer as un-opinionated as possible, sticking as close to the "Image World Models" paper's interpretation as possible. Let me know how this looks!

**Review by huguesva — COMMENTED** [review]
super cool @sami-bg  ! Did you have the occasion to try it on CIFAR10 for instance?

**Comment by huguesva** [comment]
I think you are right to break it down into several small PRs :) 
Also the collators.py file is probably a good idea, to keep things well partitioned.

**Comment by RandallBalestriero** [comment]
Yes, we need a `collator.py` and even a `sampler.py` in the long-term to handle custom sampling from any dataset, so great to have that :) That's what I have been using as well.

**Comment by sami-bg** [comment]
> super cool @sami-bg ! Did you have the occasion to try it on CIFAR10 for instance?

Thanks! I have tried pre-training with this JEPA Trainer separately (in the repo I linked) on a synthetic dataset and on SomethingSomethingV2 and pre-training is working (loss goes down), with V-JEPA. The VJEPA-specific trainer is responsible for some of the particular tricks to make things converge well (handling EMA, weight decay, etc.), but a lot of the core flow is done by the JEPA trainer.

For this PR, I tried making my own `MyCustomJEPA` class, and using that to make a simple JEPA, but after some experimenting it seems that its code would have to move out of the `examples/` directory and into somewhere within `stable_ssl`, so that we can point to the class in the `.yaml` configs. 

Furthermore, to make the `MyCustomJEPA` work, we would need to define its collate functions (such as masking the samples) and some custom forwards (e.g. masking in the ViT's `forward`) into `stable_ssl` too. At the point when we add the `collator.py`, we could get something close to I-JEPA (still a few details I'm not sure how to integrate). I was thinking perhaps we could revisit explicit testing/examples for this JEPATrainer then?

I'd be happy to get started on some PoC for the `collator.py` though! Perhaps by porting over the masking techniques in V-JEPA and I-JEPA?

**Review by huguesva — APPROVED** [review]

**Comment by huguesva** [comment]
Perfect @sami-bg ! yes 100% you can focus on this ``collator.py`` :)
